### PR TITLE
register parameters correctly in c++ MultiheadAttention

### LINKED ISSUE #41951: Parameter registration in MultiheadAttention c++ API is incorrect
## 🐛 Bug

MultiheadAttention c++ API did not register parameters under some conditions. The most important ones are 'in_proj_weight' and 'in_proj_bias'.

Here is the parameter registration phase in c++ impl:
https://github.com/pytorch/pytorch/blob/47766e648ff16d4b1175d04710caed566de14ab4/torch/csrc/api/src/nn/modules/activation.cpp#L450

Here is the parameter registration phase in python impl:
https://github.com/pytorch/pytorch/blob/75a4862f639de666c66a0db240c993918b80707f/torch/nn/modules/activation.py#L842

The problem here is that in python, the following code will register parameter automatically, but c++ version doesn't.
```
self.in_proj_bias = Parameter(torch.empty(3 * embed_dim))
```
The reason is that `Module` has a customized __setattr__ which handles the parameter/module registration automatically:
https://github.com/pytorch/pytorch/blob/eace0533985641d9c2f36e43e3de694aca886bd9/torch/nn/modules/module.py#L657

This will cause problem when people want to update parameters' value by iterating parameter that returned by calling parameters(), since some of the parameters won't be returned by c++ version. Here is an example:
https://github.com/pytorch/pytorch/blob/f71cccc457421ad220cf58914c8bc3801b072300/test/test_nn.py#L4678

The possible equivalent c++ impl will be something like this:
```
torch::MultiheadAttention model(torch::MultiheadAttentionOptions(4, 2).dropout(0.0));
{
    torch::NoGradGuard guard;
    for (auto& p : model->parameters()) {
      auto sz = p.view(-1).size(0);
      p.copy_(torch::cos(torch::arange(0, sz, tensor_options).view(p.sizes())));
    }
}
```
Currently, c++ code will return two less parameters from module->parameters() than python version due to not registering the parameter correctly.

## To Reproduce

Above c++ code piece is an example.

## Expected behavior

Python and c++ api should return same number of parameters when calling parameters().

## Environment

 ```
PyTorch version: 1.7.0a0+fced54a
Is debug build: No
CUDA used to build PyTorch: 9.2

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
CMake version: version 3.14.0

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration: 
GPU 0: Tesla M40
GPU 1: Tesla M40

Nvidia driver version: 418.126.02
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.7.6.5
/usr/local/cuda-9.2/targets/x86_64-linux/lib/libcudnn.so.7.1.4
```

## Additional context

cc @yf225 @glaringlee

### PR DESCRIPTION
**Pull request by glaringlee**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42037 register parameters correctly in c++ MultiheadAttention**

This is to fix #41951

Differential Revision: [D22764717](https://our.internmc.facebook.com/intern/diff/D22764717)

### CODE REVIEW

**Comment by glaringlee** [comment]
This is to correct the parameter registration in MultiheadAttention, the root cause is described in #41951.
cc @pbelevich @yf225

File: test/cpp/api/modules.cpp [hunk 1]
```diff
@@ -3200,6 +3201,9 @@ namespace detail {
       } else {
         std::uniform_int_distribution<int> d(5, 20);
         kv_dim = d(generator);
+        while (kv_dim == d_model) {
```
Comment (zhangguanheng66) [inline test/cpp/api/modules.cpp:3204]: what's this for?

File: torch/csrc/api/src/nn/modules/activation.cpp [hunk 2]
```diff
@@ -475,8 +480,8 @@ void MultiheadAttentionImpl::reset() {
       options.embed_dim()).bias(options.bias()))
   );
   if (options.add_bias_kv()) {
-    bias_k = torch::empty({1, 1, options.embed_dim()});
-    bias_v = torch::empty({1, 1, options.embed_dim()});
+    bias_k = register_parameter("bias_k", torch::empty({1, 1, options.embed_dim()}));
+    bias_v = register_parameter("bias_v", torch::empty({1, 1, options.embed_dim()}));
   } else {
     bias_k = {};
```
Comment (zhangguanheng66) [inline torch/csrc/api/src/nn/modules/activation.cpp:486]: Do we need to `register_parameter('bias_k', {}, /*requires_grad=*/false)`?

**Review by zhangguanheng66 — COMMENTED** [review]
Not relevant. We merged two PRs in H1 for nn.MHA on the Python side. 
https://github.com/pytorch/pytorch/pull/31996
https://github.com/pytorch/pytorch/pull/33763

File: torch/csrc/api/src/nn/modules/activation.cpp (hunk 2 above)
Comment (glaringlee) [reply torch/csrc/api/src/nn/modules/activation.cpp:486]: Based on python impl, No. It only register the non-empty case for bias_k and bias_v.

File: test/cpp/api/modules.cpp (hunk 1 above)
Comment (glaringlee) [reply test/cpp/api/modules.cpp:3204]: This branch was aimed to test kv_dim != d_model case, but the generator d could generate a same value as d_model, for eg. 9, so I add this while.

**Review by zhangguanheng66 — APPROVED** [review]
LGTM for the MHA functionality.

**Review by yf225 — APPROVED** [review]

**Pull request closed without merging**